lib/web/mfa.go

```diff
@@ -211,3 +212,161 @@ func getDeviceUsage(reqUsage string) (proto.DeviceUsage, error) {
+	// requires MFA check.
+	Database *isMFARequiredDatabase `json:"database,omitempty"`
+	// Node contains fields required to check if target node
+	// requires MFA check.
+	Node *isMFARequiredNode `json:"node,omitempty"`
+	// WindowsDesktop contains fields required to check if target
+	// windows desktop requires MFA check.
+	WindowsDesktop *isMFARequiredWindowsDesktop `json:"windows_desktop,omitempty"`
+	// Kube is the name of the kube cluster to check if target cluster
+	// requires MFA check.
+	Kube *isMFARequiredKube `json:"kube,omitempty"`
+}
+
+func (r *isMFARequiredRequest) checkAndGetProtoRequest() (*proto.IsMFARequiredRequest, error) {
+	numRequests := 0
+	var protoReq *proto.IsMFARequiredRequest
+
+	if r.Database != nil {
+		numRequests++
+		if r.Database.ServiceName == "" {
+			return nil, trace.BadParameter("missing service_name for checking database target")
+		}
+		if r.Database.Protocol == "" {
+			return nil, trace.BadParameter("missing protocol for checking database target")
+		}
+
+		protoReq = &proto.IsMFARequiredRequest{
+			Target: &proto.IsMFARequiredRequest_Database{
+				Database: &proto.RouteToDatabase{
+					ServiceName: r.Database.ServiceName,
+					Protocol:    r.Database.Protocol,
+					Database:    r.Database.DatabaseName,
+					Username:    r.Database.Username,
+				}},
+		}
+	}
+
+	if r.Kube != nil {
+		numRequests++
+		if r.Kube.ClusterName == "" {
+			return nil, trace.BadParameter("missing cluster_name for checking kubernetes cluster target")
+		}
+
+		protoReq = &proto.IsMFARequiredRequest{
+			Target: &proto.IsMFARequiredRequest_KubernetesCluster{
+				KubernetesCluster: r.Kube.ClusterName,
+			},
+		}
+	}
+
+	if r.WindowsDesktop != nil {
+		numRequests++
+		if r.WindowsDesktop.DesktopName == "" {
+			return nil, trace.BadParameter("missing desktop_name for checking windows desktop target")
+		}
+		if r.WindowsDesktop.Login == "" {
+			return nil, trace.BadParameter("missing login for checking windows desktop target")
+		}
+
+		protoReq = &proto.IsMFARequiredRequest{
+			Target: &proto.IsMFARequiredRequest_WindowsDesktop{
+				WindowsDesktop: &proto.RouteToWindowsDesktop{
+					WindowsDesktop: r.WindowsDesktop.DesktopName,
+					Login:          r.WindowsDesktop.Login,
+				}},
+		}
+	}
+
+	if r.Node != nil {
+		numRequests++
+		if r.Node.Login == "" {
+			return nil, trace.BadParameter("missing login for checking node target")
+		}
+		if r.Node.NodeName == "" {
+			return nil, trace.BadParameter("missing node_name for checking node target")
+		}
+
+		protoReq = &proto.IsMFARequiredRequest{
+			Target: &proto.IsMFARequiredRequest_Node{
+				Node: &proto.NodeLogin{
+					Login: r.Node.Login,
+					Node:  r.Node.NodeName,
+				}},
+		}
+	}
+
+	if numRequests > 1 {
+		return nil, trace.BadParameter("only one target is allowed for MFA check")
+	}
+
+	if protoReq == nil {
+		return nil, trace.BadParameter("missing target for MFA check")
+	}
+
+	return protoReq, nil
+}
+
+type isMfaRequiredResponse struct {
```

Comment:
Did we miss this one in the renames?

Reply:
OH NOOOO!!! i did... 😭 i'll fix them in the backports, for master i'll sneak it in

Reply:
No worries! Thanks for the fix.